### DISCUSSION TITLE
5. Pathlib.nvim

### DISCUSSION THREAD

**Discussion by pysan3** [nvim-neo-tree/neo-tree.nvim#1339] [Internal Development]
### 6. Pathlib.nvim

I am currently experimenting with a new way to manage path object in lua with [pathlib.nvim](https://github.com/pysan3/pathlib.nvim). I believe this library improves the overall performance regarding to path management, and should decrease the maintenance burden of dealing with path strings.
Neo-tree currently has not small amount of Windows path related issues, which should be fixed right away.

**Comment by cseickel** [comment]
Hell yeah, that looks awesome.

**Reply by pysan3** [reply]
If I introduce this library to the code, every path will be a `PathlibPath` object which will change the type annotations here and there.

Internal changes wouldn't matter that much if we be careful, but what matters the most is the event_handlers which is facing the users as the type of argument will be changed for `file_opened`, `file_moved` etc.

What do you think is the best way to go?

1. make new handlers with name appended with `pathlib`. (`file_opened` => `file_opened_pathlib`) and keep the old events as is.
2. just break it

Passing `PathlibPath` will have the added benefit that users don't need to write complicated code with `vim.loop` as this object has `path:is_dir()` and other useful methods.

**Reply by cseickel** [reply]
While I absolutely love pathlib and I think we should use it for all _internal_ comparison and generation of file paths, I do not think we should be passing it as an argument to events. In that case, it should just continue to be a simple string. It will be the string produced by pathlib though and therefor standardized in a predictable way.
